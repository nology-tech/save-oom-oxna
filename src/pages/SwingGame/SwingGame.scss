@use "../../styles/variables" as *;

.swing-game {
  background-image: url(../../assets/images/GameBackground.png);
  background-size: 100% 100%;
  height: 100%;

  .swing-game__phonic {
    display: flex;
    flex-direction: column;
    align-items: center;
    gap: 3rem;
  }

  .swing-game__squirrel {
    position: absolute;
    width: 6rem;
    height: 6rem;
    cursor: pointer;

    bottom: 5%;
    left: 30%;
    -webkit-animation-duration: 2s;
    animation-duration: 2s;
  }

  .swing-game__oom {
    position: absolute;
    width: 14rem;
    height: 14rem;
    bottom: 0.5%;
    left: 65%;
    -webkit-animation-duration: 7s;
    animation-duration: 7s;
    transform: rotate(20deg);
  }

  .swing-game__score {
    position: absolute;
    bottom: 5%;
    left: 50%;
    width: unset;
  }
}

.animate__swinging {
  transform-origin: top center;
  animation-name: swinging;
}

@media screen and (min-width: 1100px) {
  .swing-game {
    .swing-game__squirrel {
      width: 10rem;
      height: 10rem;
    }
    .swing-game__oom {
      width: 18rem;
      height: 18rem;
    }
  }
}

@keyframes swinging {
  0% {
    transform: rotate(20deg);
  }
  50% {
    transform: rotate(-15deg);
  }
  75% {
    transform: rotate(50deg);
  }
  100% {
    transform: rotate(-15deg);
  }
}

.animate__animated.animate__fastest {
  -webkit-animation-duration: calc(1s * 0.01);
  animation-duration: calc(1s * 0.01);
  -webkit-animation-duration: calc(var(--animate-duration) * 0.01);
  animation-duration: calc(var(--animate-duration) * 0.01);
}
