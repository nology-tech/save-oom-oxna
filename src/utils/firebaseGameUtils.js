import { db } from "../firebase";
import {
  addDoc,
  collection,
  doc,
  getDocs,
  getDoc,
  query,
  where,
  orderBy,
  Timestamp,
} from "firebase/firestore";


/**
 * Gets a user object using the userId.
 * @param {*} userId 
 * @returns 
 */
export const getUserById = async( userId ) => {
  const docRef = doc(db, "users", userId);
  const docSnap = await getDoc( docRef );
  return docSnap;
}

/** 
 * Get all the rounds of a specific game for a given user, ordered by most recent first 
 @param {*} userId
*@returns a querySnap (document snapshot), returns a Promise.
*/

export const getGameRoundsForUser = async (id, game) => {
  const q = query(
    collection(db, "users", id, "rounds_played"),
    where("gameId", "==", game),
    orderBy("roundTime", "desc")
  );
  const querySnap = await getDocs(q);
  return querySnap;
};

/** 
   * Get all the correct round of a specific game for a given user, ordered by most recent first 
  @param {String} userId
  @param {String} gameId
  *@returns a promise with the query snapshot.
  */
export const getCorrectGameRoundsForUser = async (id, game) => {
  const q = query(
    collection(db, "users", id, "rounds_played"),
    where("gameId", "==", game),
    where("correct", "==", true),
    orderBy("roundTime", "desc")
  );
  const querySnap = await getDocs(q);
  return querySnap;
};

/** Get all the incorrect round of a specific game for a given user, ordered by most recent first 
  @param {String} userId
  @param {String} gameId
  *@returns  a promise with the query snapshot.
  */
export const getIncorrectGameRoundsForUser = async (id, game) => {
  const q = query(
    collection(db, "users", id, "rounds_played"),
    where("gameId", "==", game),
    where("correct", "==", false),
    orderBy("roundTime", "desc")
  );
  const querySnap = await getDocs(q);
  return querySnap;
};

/** Function that returns an array of rounds 
  @param {String} userId
  @param {String} gameId
  @param {Function} getGameRounds
  *@returns returns a promise.
  */
export const getArrayOfRounds = async (id, game, getGameRounds) => {
  const querySnap = await getGameRounds(id, game);
  const phonicsArr = querySnap.docs.map((doc) => {
    return doc.data().phonic;
  });
  return phonicsArr;
};

/** Function that returns an array of rounds 
  @param {String} id
  @param {String} game
  @param {Object} roundInfo
  @param {String} answer
  *@returns returs a promise.
  */
export const saveUserRound = async (id, game, level, gameStats, answer) => {
  const { isCorrect, score } = gameStats;
  await addDoc(collection(db, "users", id, "rounds_played"), {
    correct: isCorrect,
    gameId: game,
    phonic: answer,
    level: level,
    roundTime: Timestamp.fromDate(new Date()).toDate(),
    score: score,
  });
};
