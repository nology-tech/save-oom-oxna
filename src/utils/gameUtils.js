import phonicsData from "../data/phonicsData";
import { 
    // getGameRoundsForUser, 
    getCorrectGameRoundsForUser, 
    getIncorrectGameRoundsForUser,
    getArrayOfRounds
     } from "./firebaseGameUtils";

// const staticPhonicsArray = Object.keys(phonicsData.levelOne);

/**
 * Get a list of phonics from the static array, based on the level.
 * @param {*} level 
 * @returns 
 */
const getStaticPhonicsArray = (level) => {
    let levelData = null;
    switch(level) {
        case 1:
            levelData = "levelOne";
            break;
        case 2:
            levelData = "levelTwo";
            break;
        case 3:
            levelData = "levelThree";
            break;
        case 4:
            levelData = "levelFour";
            break;
        case 5:
            levelData = "levelFive";
            break;
    }
    return Object.keys( phonicsData[levelData] );
}

/**
 * Shuffle that array. Nice and simple algorithm.
 * Code courtesy of Geeks for Geeks.
 * @param {*} array 
 * @returns 
 */
const shuffleArray = (array) => {
    for (var i = array.length - 1; i > 0; i--) {
    
        // Generate random number
        var j = Math.floor(Math.random() * (i + 1));
                    
        var temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
        
    return array;
}
    

/**
 * Get the array of sounds for the swing game.
 * Attempts to return a mix of incorrect, correct and new sounds for the given child, based on their
 * past results. Basic algo at the moment!
 * @param {*} id 
 * @param {*} level 
 * @returns 
 */
export const getArrayForSwing = async (id, level) => {

    // 50 wrong = 8
    // 30 new = 4
    // fill up from rest = 3

    const MAX_PER_GAME = 15;
    const MAX_INCORRECT_PER_GAME = 8;
    const MAX_CORRECT_PER_GAME = 4;

    const game = "swing";

    const staticArray = getStaticPhonicsArray( level );
    console.log(`static array for level ${level}`, staticArray);

    const incorrect = await getArrayOfRounds( id, game, getIncorrectGameRoundsForUser);
    const correct = await getArrayOfRounds( id, game, getCorrectGameRoundsForUser);

    console.log( "correct", correct);
    console.log( "incorrect", incorrect);

    // all results
    const result = new Set();

    // add in at most MAX_INCORRECT_PER_GAME of incorrect answers
    for (let phonic of incorrect) {
        result.add(phonic);
        if( result.length >= MAX_INCORRECT_PER_GAME ) {
            console.log("Break!! Reach max incorrect");
            break;
        }
    }
    console.log("After adding incorrect", result);

    let maxCorrect = result.length + MAX_CORRECT_PER_GAME;
    // add in at most MAX_CORRECT_PER_GAME of correct answers
    for (let phonic of correct) {
        result.add(phonic);
        if( result.length >= maxCorrect ) {
            break;
        }
    }
    console.log("After adding correct", result);

    // fill up remaining
    for (let i = 0; i < staticArray.length; i++) {
        result.add( staticArray[i] );
        if( result.size >= MAX_PER_GAME ) {
            break;
        }
    }

    console.log("At end", result);

    const shuffled = shuffleArray( Array.from(result) );

    return shuffled;
}