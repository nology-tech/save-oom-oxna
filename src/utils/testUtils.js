import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import UserContext from "../contexts/UserContext";

export const customRender = (
  ui,
  useRouting = true,
  contextOptions = { useContext: true, value: undefined }
) => {
  // wrap components in routing and context if requested
  const routingUiResult = useRouting && wrapWithRouting(ui);
  const contextUiResult =
    contextOptions.useContext &&
    wrapWithContext(routingUiResult, contextOptions.value);

  // use RTL's render function to return the test component
  return render(contextUiResult);
};

const wrapWithRouting = (ui) => {
  return <Router>{ui}</Router>;
};

const wrapWithContext = (ui, value) => {
  return <UserContext.Provider value={value}>{ui}</UserContext.Provider>;
};
