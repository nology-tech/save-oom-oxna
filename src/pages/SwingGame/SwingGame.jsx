import { useContext, useEffect, useState } from "react";
import { saveUserRound } from "../../api/gameService";
import swingingOom from "../../assets/images/Group 146swingingOom.png";
import squirrel from "../../assets/images/squirrel.png";
import AnimatedImage from "../../components/AnimatedImage/AnimatedImage";
import ButtonRoundContainer from "../../components/ButtonRoundContainer/ButtonRoundContainer";
import Layout from "../../components/Layout/Layout";
import NeedsContainer from "../../components/NeedsContainer/NeedsContainer";
import PhonicComponent from "../../components/PhonicComponent/PhonicComponent";
import Timer from "../../components/Timer/Timer";
import UserContext from "../../context/UserContext";
import phonicsData from "../../data/phonicsData";
import { getArrayForSwing } from "../../utils/gameUtils";
import GameEnd from "../GameEnd/GameEnd";
import "./SwingGame.scss";

const SwingGamePage = () => {
  const [gameState, setGameState] = useState({
    score: 0,
    index: 0,
    isCorrect: false,
  });
  const [hintAnimation, setHintAnimation] = useState("");
  const [oomAnimation, setOomAnimation] = useState("");
  const [phonicsArray, setPhonicsArray] = useState([]);
  const [isGameOver, setIsGameOver] = useState(false);
  const [isLoaded, setIsLoaded] = useState(false);

  const { user } = useContext(UserContext);

  // load up the phonics array
  useEffect(async () => {
    if (!user.userId) return;
    const array = await getArrayForSwing(user.userId, 1);
    setPhonicsArray(array);
    setIsLoaded(true);
  }, [user.userId]);

  // adds animate class to squirrel, removes after 3 seconds
  const triggerHintAnimation = () => {
    setHintAnimation("animate__bounce");
    setInterval(() => {
      setHintAnimation("");
    }, 3000);
  };

  // adds animate class to oom, removes after 8 seconds
  const triggerOomAnimation = () => {
    setOomAnimation("animate__swinging");
    setInterval(() => {
      setOomAnimation("");
    }, 8000);
  };

  const handleAnswer = async (isCorrect) => {
    setHintAnimation(false);
    if (isCorrect) {
      triggerOomAnimation();
    }
    const currentPhonic = phonicsArray[gameState.index];
    const score = gameState.score + (isCorrect ? 1 : 0);
    const index = handleIndexChange();

    setGameState({
      score,
      isCorrect,
      index,
    });

    try {
      await saveUserRound(
        user.userId,
        "swing",
        1,
        { isCorrect, score },
        currentPhonic
      );
      console.log(`Saved user round - ${isCorrect ? "" : "in"}correct!`);
    } catch (e) {
      console.error("Error attempting to save a user's round");
    }
  };

  // TODO: stop sound playing multiple times
  const handleHint = async () => {
    const currentPhonic = phonicsArray[gameState.index];
    const phonicSound = await phonicsData.levelOne[currentPhonic].sound();
    new Audio(phonicSound).play();

    triggerHintAnimation();
  };

  const handleIndexChange = () => {
    return gameState.index < phonicsArray.length ? gameState.index + 1 : 0;
  };

  const handleGameEnd = () => {
    setIsGameOver(true);
  };

  if (!isLoaded) {
    return (
      <Layout>
        <div className="swing-game">
          <p>Loading...</p>
        </div>
      </Layout>
    );
  }

  if (isGameOver) {
    return (
      <Layout>
        <div className="swing-game">
          <GameEnd score={gameState.score} childName={user.name} />
        </div>
      </Layout>
    );
  }

  return (
    <Layout>
      <div className="swing-game">
        <NeedsContainer />
        <div className="swing-game__phonic">
          <Timer startTime={6000} onTimeUp={handleGameEnd} />
          <PhonicComponent phonicText={phonicsArray[gameState.index]} />
        </div>
        <div onClick={handleHint}>
          <AnimatedImage
            imageToAnimate={squirrel}
            animationClass="animate__animated.animate__fastest"
            animationType={hintAnimation}
            imageStylesClass="swing-game__squirrel"
          />
        </div>
        <AnimatedImage
          imageToAnimate={swingingOom}
          animationClass="animate__animated.animate__fastest"
          animationType={oomAnimation}
          imageStylesClass="swing-game__oom"
        />
        <ButtonRoundContainer
          handleCorrect={() => handleAnswer(true)}
          handleIncorrect={() => handleAnswer(false)}
        />
        <p className="swing-game__score">
          Number Of Correct Sounds: {gameState.score}
        </p>
      </div>
    </Layout>
  );
};

export default SwingGamePage;
