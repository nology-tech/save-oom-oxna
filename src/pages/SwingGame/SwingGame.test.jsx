import { customRender } from "../../utils/testUtils";
import SwingGame from "./SwingGame";

it("Should render the SwingGame page", () => {
  const { container } = customRender(<SwingGame />);

  expect(container).toMatchSnapshot();
});
