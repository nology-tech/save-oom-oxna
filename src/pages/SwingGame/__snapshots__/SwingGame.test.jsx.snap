// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the SwingGame page 1`] = `
<div>
  <div
    class="layout"
  >
    <nav
      class="nav-bar"
    >
      <div>
        <a
          href="/dashboard"
        >
          <img
            alt="logo"
            class="logo"
            src="logo.png"
          />
        </a>
        <a
          class="navbar-link"
          href="/swing-game"
        >
          Swing Game
        </a>
        <a
          class="navbar-link"
          href="/dashboard"
        >
          Dashboard
        </a>
        <a
          class="navbar-link"
          href="/stats-page"
        >
          Stats
        </a>
        <h2>
          Barry
        </h2>
      </div>
      <button
        class="button button--secondary"
      >
        Logout
      </button>
    </nav>
    <main
      class="layout__main"
    >
      <div
        class="swing-game"
      >
        <p>
          Loading...
        </p>
      </div>
    </main>
  </div>
</div>
`;
